File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/ddd/shared/Quantity.java
package pe.joedayz.microservicios.modulo01.ddd.shared;

/**
 * Value Object: cantidad de unidades. Siempre positiva (invariante del dominio).
 */
public record Quantity(int value) {

    public Quantity {
        if (value <= 0) {
            throw new IllegalArgumentException("La cantidad debe ser mayor a 0, recibido: " + value);
        }
    }

    public static Quantity of(int value) {
        return new Quantity(value);
    }

    @Override
    public String toString() {
        return Integer.toString(value);
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Ddd/Shared/Quantity.cs
namespace JoeDayz.Microservicios.Modulo01.Ddd.Shared;

/// <summary>
/// Cantidad siempre positiva. Un value object evita que un <c>int</c> suelto
/// se cuele con valores imposibles (0 o negativos) en el dominio.
/// </summary>
public readonly record struct Quantity
{
    public Quantity(int value)
    {
        if (value <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(value), value, "La cantidad debe ser mayor a 0");
        }

        Value = value;
    }

    public int Value { get; }

    public static Quantity Of(int value) => new(value);

    public override string ToString() => Value.ToString();
}
